File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/box/GroundFrictionForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.box;

import com.github.aadvorak.artilleryonline.battle.calculations.BodyForce;
import com.github.aadvorak.artilleryonline.battle.calculations.BoxCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.ForceCalculator;
import com.github.aadvorak.artilleryonline.battle.calculator.utils.ForceCalculatorUtils;
import com.github.aadvorak.artilleryonline.battle.config.BoxConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.BoxModel;
import com.github.aadvorak.artilleryonline.battle.precalc.BoxPreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.BoxSpecs;
import com.github.aadvorak.artilleryonline.battle.state.BoxState;

import java.util.ArrayList;
import java.util.List;

public class GroundFrictionForceCalculator implements ForceCalculator<
        BoxSpecs,
        BoxPreCalc,
        BoxConfig,
        BoxState,
        BoxModel,
        BoxCalculations> {

    private static final String FORCE_DESCRIPTION = "Ground Friction";

    @Override
    public List<BodyForce> calculate(BoxCalculations calculations, BattleModel battleModel) {
        var groundFrictionCoefficient = battleModel.getRoom().getSpecs().getGroundFrictionCoefficient();
        var gravityAcceleration = battleModel.getRoom().getSpecs().getGravityAcceleration();
        var contactsNumber = calculations.getGroundContacts().size();
        var forces = new ArrayList<BodyForce>();
        if (calculations.getGroundContacts() != null) {
            var forceMultiplier = groundFrictionCoefficient * gravityAcceleration
                    * calculations.getMass() / contactsNumber;
            calculations.getGroundContacts().forEach(contact -> ForceCalculatorUtils.addFriction(forces,
                    calculations, contact, forceMultiplier, FORCE_DESCRIPTION));
        }
        return forces;
    }
}

File: frontend/playground/battle/calculator/common/ground-friction-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import {type BodyCalculations} from "~/playground/data/calculations";
import type {BattleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {BodyUtils} from "~/playground/utils/body-utils";

export class GroundFrictionForceCalculator implements ForceCalculator<BodyCalculations> {
  private static readonly FORCE_DESCRIPTION = 'Ground Friction'

  calculate(calculations: BodyCalculations, battleModel: BattleModel): BodyForce[] {
    const groundFrictionCoefficient = battleModel.room.specs.groundFrictionCoefficient
    const gravityAcceleration = battleModel.room.specs.gravityAcceleration
    const contactsNumber = BodyUtils.getAllGroundContacts(calculations).length
    const forces: BodyForce[] = []
    this.addFriction(forces, calculations, groundFrictionCoefficient, gravityAcceleration, contactsNumber)
    return forces
  }

  private addFriction(
      forces: BodyForce[],
      calculations: BodyCalculations,
      groundFrictionCoefficient: number,
      gravityAcceleration: number,
      contactsNumber: number
  ): void {
    const groundContacts = calculations.groundContacts
    if (!groundContacts) {
      return
    }

    const forceMultiplier = groundFrictionCoefficient * gravityAcceleration
        * calculations.model.preCalc.mass / contactsNumber

    groundContacts.forEach(contact => {
      const tangential = VectorUtils.tangential(contact.angle)
      const movingVelocity = VectorUtils.projectionOfOnto(calculations.model.state.velocity, tangential)
      const rotatingVelocity = VectorUtils.projectionOfOnto(
          BodyUtils.getRotatingVelocityAt(calculations.model.state, contact.position), tangential
      )
      const movingVelocityMagnitude = VectorUtils.getMagnitude(movingVelocity)
      const rotatingVelocityMagnitude = VectorUtils.getMagnitude(rotatingVelocity)
      const contactForceMultiplier = forceMultiplier * Math.cos(contact.angle)

      if (movingVelocityMagnitude > rotatingVelocityMagnitude) {
        const velocity = BodyUtils.getVelocityAt(calculations.model.state, contact.position)
        const movingForce = {
          x: -velocity.x * contactForceMultiplier,
          y: -velocity.y * contactForceMultiplier
        }
        forces.push(BodyForce.of(
            movingForce,
            contact.position,
            calculations.model.state.position,
            GroundFrictionForceCalculator.FORCE_DESCRIPTION
        ))
      } else {
        const rotatingForce = {
          x: -rotatingVelocity.x * contactForceMultiplier,
          y: -rotatingVelocity.y * contactForceMultiplier
        }
        forces.push(BodyForce.atCOM(rotatingForce, GroundFrictionForceCalculator.FORCE_DESCRIPTION))
      }
    })
  }
}
